File: app/components/folio/console/current_users/preference_toggle_component.rb
# frozen_string_literal: true

class Folio::Console::CurrentUsers::PreferenceToggleComponent < Folio::Console::ApplicationComponent
  def initialize(key:, label:, javascript_key:, class_name: nil)
    @key = key
    @label = label
    @javascript_key = javascript_key
    @class_name = class_name
  end

  def enabled
    return @enabled unless @enabled.nil?

    @enabled = if Folio::Current.user && Folio::Current.user.console_preferences.present?
      Folio::Current.user.console_preferences[@key]
    end

    @enabled = true if @enabled.nil?

    @enabled
  end

  def data
    stimulus_controller("f-c-current-users-preference-toggle",
                        values: {
                          api_url: controller.update_console_preferences_console_api_current_user_path,
                          enabled:,
                          key: @key,
                          javascript_key: @javascript_key,
                        },
                        action: {
                          "f-c-ui-boolean-toggle:input" => "booleanToggleInput",
                        })
  end
end

File: app/components/folio/console/current_users/preference_toggle_component.js
window.Folio.Stimulus.register('f-c-current-users-preference-toggle', class extends window.Stimulus.Controller {
  static values = {
    enabled: { type: Number, default: -1 },
    apiUrl: String,
    key: String,
    javascriptKey: String
  }

  connect () {
    window.FolioConsole[this.javascriptKeyValue] = window.FolioConsole[this.javascriptKeyValue] || {}
    window.FolioConsole[this.javascriptKeyValue].enabled = this.enabledValue === 1
  }

  enabledValueChanged (value, previousValue) {
    if (value === previousValue) return

    window.FolioConsole[this.javascriptKeyValue] = window.FolioConsole[this.javascriptKeyValue] || {}
    window.FolioConsole[this.javascriptKeyValue].enabled = this.enabledValue === 1

    if (previousValue === -1) return

    if (window.FolioConsole[this.javascriptKeyValue].enabledChangeCallback) {
      window.FolioConsole[this.javascriptKeyValue].enabledChangeCallback()
    }

    const data = {}
    data[this.keyValue] = this.enabledValue === 1

    window.Folio.Api.apiPost(this.apiUrlValue, data)
  }

  booleanToggleInput (e) {
    this.enabledValue = e.detail.checked ? 1 : 0
  }

  stopPropagation (e) {
    e.stopPropagation()
  }
})
